Pokemon App/Pokemon App/CardListCollectionView/CardListCollectionView.swift

```diff
@@ -23,15 +23,21 @@ extension CardListCollectionView: UICollectionViewDataSource {
         
         let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "CardCollectionViewCell", for: indexPath) as! CardCollectionViewCell
         
-        let pokemonCard = pokemonList[indexPath.row]
-        
-        cell.downloadImage(imageURL: pokemonCard.imageURL)
-        
-        
         return cell
         
     }
     
+    func collectionView(_ collectionView: UICollectionView, willDisplay cell: UICollectionViewCell, forItemAt indexPath: IndexPath) {
+        let pokemonCard = pokemonList[indexPath.row]
+        let cell = cell as! CardCollectionViewCell
+        if let image = pokemonCard.image {
+            cell.pokemonCardImageView.image = image
+        } else {
+            cell.pokemonImageURL = pokemonCard.imageURL
+            cell.downloadImage()
+        }
```

Comment:
Nice ini implementasi in memory cache. tapi perlu diperhatikan juga, ini bisa membengkak RAMnya. jadi perlu dibikin `pokemonCard.image = nil` untuk deallocated unused image dari memory. Caranya bisa bikin threshold untuk 40+ item dengan index yang sudah tidak keliahatan (misalnya user udah scroll ke index 100 misalnya), nah kemungkinan index 0-40 ini sudah gadipakai lagi jadi bisa dibuang imagenya dari memory, buangnya dimana? bisa di func didEndDisplaying atau di `func didreceivememorywarning` -> https://developer.apple.com/documentation/uikit/uiviewcontroller/1621409-didreceivememorywarning . 
atau bisa Cara lain bisa disimpan di local storage exp: core data. Trade-off klo simpan dilocal storage gak secepat simpan diram.

Reply:
mas, nggak ngerti per-prefetch-an ini, misal kita di index 100, yang dibuang 0-40? jadi pokemon nya di loop terus yg index 0-40 dijadiin nil image nya? terus di index 101 dst ulang lagi gitu?._.

btw mas, kok misal aku scroll ke atas, pokemonnya download ulang image nya ya? tidak paham._.

Reply:
core data bukannya tiket selajutnya mas? jadi ini tuh pilih satu pake pre-fetch atau core data aja atau dua-duanya?